### PR TITLE
Handlers

### PR DESCRIPTION
**Pull request by 2Cubed**
New handler system! :tada:

### CODE REVIEW

**Comment by Innectic** [comment]
Woo!

**Comment by Innectic** [comment]
Aw, travis is mad

**Comment by 2Cubed** [comment]
Getting ready to merge this soon. We need to all review it, though, and make sure everything's good!

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file cactusbot/packets/event.py, line 7.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/packets/event.py
+++ b/cactusbot/packets/event.py
@@ -4,7 +4,7 @@
 
 
 class EventPacket(Packet):
-    
+
     TYPE = "event"
 
     def __init__(self, event_type, user, success=True):
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file config.template.py, line 13.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/config.template.py
+++ b/config.template.py
@@ -10,6 +10,7 @@
 PASSWORD = "p455w0rd"
 
 CHANNEL = "ChannelName"
-handlers = Handlers(LoggingHandler(), EventHandler(), CommandHandler(CHANNEL), SpamHandler())
+handlers = Handlers(LoggingHandler(), EventHandler(),
+                    CommandHandler(CHANNEL), SpamHandler())
 
 SERVICE = BeamHandler(CHANNEL, handlers)
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file cactusbot/packet.py, line 4.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/packet.py
+++ b/cactusbot/packet.py
@@ -2,6 +2,7 @@
 
 
 class Packet:
+
     def __init__(self, packet_type, **kwargs):
         self.TYPE = packet_type
         self.kwargs = kwargs
```

File: cactusbot/services/beam/constellation.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+"""Interact with Beam Constellation."""
+
+import re
+import json
+
+from .. import WebSocket
```
Comment (Innectic) [inline cactusbot/services/beam/constellation.py]: The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/services/beam/constellation.py
+++ b/cactusbot/services/beam/constellation.py
@@ -10,7 +10,7 @@
     """Interact with Beam Constellation."""
 
     URL = "wss://constellation.beam.pro"
-    
+
     RESPONSE_EXPR = re.compile(r'^(\d+)(.+)?$')
     INTERFACE_EXPR = re.compile(r'^([a-z]+):\d+:([a-z]+)')
 
```

**Comment by Innectic** [comment]
Comment on 3981325e018d132c9385f30cb0a87488fd28a7a6, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

**Comment by Innectic** [comment]
Comment on 52b630e6111181831bc4ad8ba9111481e618b4f4, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

File: cactusbot/commands/magic/cactus.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        return MessagePacket(
+            ("text", "Check out my documentation at "),
+            ("link", "https://cactusbot.rtfd.org", "cactusbot.rtfd.org"),
+            ("text", ".")
+        )
+
+    @Command.subcommand
+    async def help(self):
+        return "Try our docs (!cactus docs). If that doesn't help, tweet at us (!cactus twitter)!"
+
```
Comment (Innectic) [inline cactusbot/commands/magic/cactus.py]: The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/commands/magic/cactus.py
+++ b/cactusbot/commands/magic/cactus.py
@@ -27,7 +27,7 @@
     @Command.subcommand
     async def help(self):
         return "Try our docs (!cactus docs). If that doesn't help, tweet at us (!cactus twitter)!"
-    
+
     @Command.subcommand()
     async def twitter(self):
         return MessagePacket(
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file cactusbot/packets/event.py, line 7.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/packets/event.py
+++ b/cactusbot/packets/event.py
@@ -4,7 +4,7 @@
 
 
 class EventPacket(Packet):
-    
+
     TYPE = "event"
 
     def __init__(self, event_type, user, success=True):
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file cactusbot/packet.py, line 4.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/packet.py
+++ b/cactusbot/packet.py
@@ -2,6 +2,7 @@
 
 
 class Packet:
+
     def __init__(self, packet_type, **kwargs):
         self.TYPE = packet_type
         self.kwargs = kwargs
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file config.template.py, line 13.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/config.template.py
+++ b/config.template.py
@@ -10,6 +10,7 @@
 PASSWORD = "p455w0rd"
 
 CHANNEL = "ChannelName"
-handlers = Handlers(LoggingHandler(), EventHandler(), CommandHandler(CHANNEL), SpamHandler())
+handlers = Handlers(LoggingHandler(), EventHandler(),
+                    CommandHandler(CHANNEL), SpamHandler())
 
 SERVICE = BeamHandler(CHANNEL, handlers)
```

**Comment by Innectic** [comment]
Comment on 0b0bc1f527e23ddf98dfe85e5635afc14664f607, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

File: cactusbot/services/beam/constellation.py (hunk 1 above)
Comment (Innectic) [inline cactusbot/services/beam/constellation.py]: The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/services/beam/constellation.py
+++ b/cactusbot/services/beam/constellation.py
@@ -10,7 +10,7 @@
     """Interact with Beam Constellation."""
 
     URL = "wss://constellation.beam.pro"
-    
+
     RESPONSE_EXPR = re.compile(r'^(\d+)(.+)?$')
     INTERFACE_EXPR = re.compile(r'^([a-z]+):\d+:([a-z]+)')
 
```

**Comment by Innectic** [comment]
Comment on 3981325e018d132c9385f30cb0a87488fd28a7a6, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

**Comment by Innectic** [comment]
Comment on 52b630e6111181831bc4ad8ba9111481e618b4f4, file docs/conf.py, line 214.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -211,17 +211,17 @@
 # -- Options for LaTeX output ---------------------------------------------
 
 latex_elements = {
-# The paper size ('letterpaper' or 'a4paper').
-#'papersize': 'letterpaper',
-
-# The font size ('10pt', '11pt' or '12pt').
-#'pointsize': '10pt',
-
-# Additional stuff for the LaTeX preamble.
-#'preamble': '',
-
-# Latex figure (float) alignment
-#'figure_align': 'htbp',
+    # The paper size ('letterpaper' or 'a4paper').
+    #'papersize': 'letterpaper',
+
+    # The font size ('10pt', '11pt' or '12pt').
+    #'pointsize': '10pt',
+
+    # Additional stuff for the LaTeX preamble.
+    #'preamble': '',
+
+    # Latex figure (float) alignment
+    #'figure_align': 'htbp',
 }
 
 # Grouping the document tree into LaTeX files. List of tuples
```

File: cactusbot/commands/magic/cactus.py (hunk 2 above)
Comment (Innectic) [inline cactusbot/commands/magic/cactus.py]: The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `default`.*

The issue can be fixed by applying the following patch:

```diff
--- a/cactusbot/commands/magic/cactus.py
+++ b/cactusbot/commands/magic/cactus.py
@@ -27,7 +27,7 @@
     @Command.subcommand
     async def help(self):
         return "Try our docs (!cactus docs). If that doesn't help, tweet at us (!cactus twitter)!"
-    
+
     @Command.subcommand()
     async def twitter(self):
         return MessagePacket(
```

File: cactusbot/commands/magic/friend.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+"""Manage friends."""
+
+from . import Command
+
+
+# TODO: Finish this command.
+class Friend(Command):
```
Comment (2Cubed) [inline cactusbot/commands/magic/friend.py]: Does not currently do anything. However, this requires the API as well, and can wait.

File: cactusbot/commands/magic/social.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+"""Get social data."""
+
+import aiohttp
+
+from . import Command
+
+from ...packets import MessagePacket
+
+
+class Social(Command):
```
Comment (2Cubed) [inline cactusbot/commands/magic/social.py]: This is Beam-specific, not service-agnostic. However, that will be fixed with #60.

File: cactusbot/handlers/command.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        if len(packet) > 1 and packet[0] == "!" and packet[1] != ' ':
+            command, *args = packet[1:].split()
+            if command in self.MAGICS:
+                response = await self.MAGICS[command](
+                    *args, username=packet.user, channel=self.channel)
+                if packet.target:
+                    response.target = packet.user
+                return response
+            else:
+                # TODO: custom commands
```
Comment (2Cubed) [inline cactusbot/handlers/command.py]: Needs API, but can wait.

File: cactusbot/handlers/spam.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+"""Handle incoming spam messages."""
+
+from ..handler import Handler
+from ..packets import BanPacket, MessagePacket
+
+
+class SpamHandler(Handler):
```
Comment (2Cubed) [inline cactusbot/handlers/spam.py]: Very happy with how this turned out - infinitely better than what we had in v0.3. :tada:

**Comment by Innectic** [comment]
MEERRRRGGGGEEE

**Pull request merged**